### DISCUSSION TITLE
[Question]: What is the Easiest Way to Deploy LibraChat to Microsoft Azure (Using a Containerized Approach and Parameter Configuration)

### DISCUSSION THREAD

**Discussion by jeannotdamoiseaux** [danny-avila/LibreChat#5012] [Q&A]
### What is your question?

What is the easiest way to deploy LibraChat to Microsoft Azure using a containerized approach (e.g., Docker) or a similar method, with configuration managed through a single parameter file? How can this be achieved with minimal setup and integration with Azure services like App Service, Kubernetes Service, or Azure Container Instances (ACI)? The deployment should also support scalability, secure connections, and compatibility with Azure's monitoring and resource management tools.

### More Details

I am looking for a simple, straightforward solution to deploy LibraChat to Azure, ideally through a containerized approach. The solution should allow me to define all necessary configuration settings (such as database credentials, storage configurations, and app-specific settings) in a single configuration file. It should also integrate easily with Azure services, ensuring scalability, security (using Azure Key Vault for sensitive data), and compatibility with Azure's monitoring features like Application Insights.

There are some existing resources I have looked at:

- The [LibreChat Bicep repo](https://github.com/jeannotdamoiseaux/ElumenotionSite/tree/master/Projects/LibreChatBicep), which provides a basic setup, but it is outdated and lacks sufficient documentation.

- Another repository found at [this link](https://github.com/thunderbug1/LibreChatAzureDeployment) also contains an implementation, but it hasn't worked for me so far and needs further adjustments.

These resources might be helpful starting points, but they require updates and improvements to be fully functional.

### What is the main subject of your question?

Hosting LibraChat on Microsoft Azure

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by jmaddington** [comment]
Getting it running on a VPS with a .env file with docker is painless. I've been running it this way for months. 

The difficulty in the other Azure services are the other images that LC depends on. Last I checked, only Kubernetes supported multiple docker images, and that takes a lot of set up. The other options don't support docker-compose, at least not outside of beta, and you'll run into persistent storage issues without a customized deployment. 

LC is very lightweight, I run it on a VPS with 2 vCPUs and 8gb RAM behind Cloudflare.

**Reply by jeannotdamoiseaux** [reply]
Thanks for your response, @jmaddington! I’m curious, how exactly does your setup work? Are you using an Azure App Container to run LibraChat, or are you strictly working outside of Azure?

I’ve already implemented deployment via Azure App Container, but my main challenge lies in setting up the infrastructure-as-code to establish a direct connection with Microsoft Entra (formerly Azure AD). Do you have any insights or advice on how to handle this integration?

**Reply by jeannotdamoiseaux** [reply]
There is a version available through the [Azure Marketplace](https://azuremarketplace.microsoft.com/nl-nl/marketplace/apps/cloud-infrastructure-services.librechat?tab=overview). However, that this version currently lacks support for Microsoft Entra integration
![image](https://github.com/user-attachments/assets/8fbfceae-8dec-40b9-be63-b936b433807f)

**Reply by jmaddington** [reply]
We have an Azure Virtual Machine setup with Docker running on it. Setting up Oauth2 login is done through the env file, I'd be glad to share with you what we have over email or DM, it was easier than I anticipated.

We are not likely to go the App Container route because of my experience with a different LOB app. Short version: Azure IAAS doesn't play nicely with multiple services unless you set up Kubernetes, which is overkill. LC requires its main service, mongo and pgvector, as well as local storage unless you use Firebase. App containers local storage is purely ephemeral, so every time you restart you lose an upload files (by design).

Anyway, setting up docker on an Ubuntu VM was painless.

If you figured out the multiples services and ephemeral storage on app containers I'd love to see what you did.

**Reply by jeannotdamoiseaux** [reply]
I would be very interested and have reached out to you on LinkedIn.

**Reply by dmki** [reply]
You can use ACI, of course. I wouldn't call it "easy", thanks to ingenious architecture of librechat, which requires separate database container. First, you need to create a docker image with the database server. Upload to Azure and run it in ACI. Then notice its non-public IP address and use it when creating second image, for your server. Upload server image, run in ACI in the same network group. 
In other words - fukitol, it was never meant for non-hobbyist deployment.

**Comment by larryb821** [comment]
I looked at azure and decided it wasn’t worth it for me (not an it pro). We host on https://elest.io/open-source/librechat and restricted login via SSO / openid connected to MS entra. Working well so far. Monthly hosting cost is small.

**Comment by jmaddington** [comment]
Here is an abridged version of my docker-compose:

```
  caddy:
    image: caddy
    
    networks:
      - librechat
      - caddy
    
    ports:
      - 80:80
      - 443:443
      - 443:443/udp
    
    restart: always

  api:
    image: <my-private-container-registry>
    container_name: LibreChat-API
    networks:
      - librechat

    depends_on:
      - mongodb
      - rag_api

    restart: always
    env_file:
      - ./LibreChat/data/.env

    environment:
      - HOST=0.0.0.0
      - NODE_ENV=production
      - MONGO_URI=mongodb://mongodb:27017/LibreChat
      - MEILI_HOST=http://meilisearch:7700
      - RAG_PORT=${RAG_PORT:-8000}
      - RAG_API_URL=http://rag_api:${RAG_PORT:-8000}

    volumes:
      - type: bind
        source: ./LibreChat/data/librechat.yaml
        target: /app/librechat.yaml
      - ./LibreChat/LibreChat/images:/app/client/public/images
      - ./LibreChat/logs:/app/api/logs

  mongodb:
    container_name: chat-mongodb
    networks:
      - librechat
    image: mongo
    restart: always
    volumes:
      - ./LibreChat/data/mongodb:/data/db
    command: mongod --noauth

 meilisearch:
    container_name: chat-meilisearch
    image: getmeili/meilisearch:v1.7.3
    restart: always
    networks:
      - librechat
    env_file:
      - ./LibreChat/data/.env
    environment:
      - MEILI_HOST=http://meilisearch:7700
      - MEILI_NO_ANALYTICS=true
    volumes:
      - ./LibreChat/data/meili_data_v1.7:/meili_data
    labels:
      - com.centurylinklabs.watchtower.enable=true

  vectordb:
    image: ankane/pgvector:latest
    env_file:
      - ./LibreChat/data/.env
    restart: always
    networks:
      - librechat
    volumes:
      - ./LibreChat/data/pg:/var/lib/postgresql/data

  rag_api:
    image: ghcr.io/danny-avila/librechat-rag-api-dev-lite:latest
    environment:
      - DB_HOST=vectordb
      - RAG_PORT=${RAG_PORT:-8000}
    networks:
      - librechat
    restart: always
    depends_on:
      - vectordb
      
    env_file:
      - ./LibreChat/data/.env

networks:
  librechat:
  caddy:
    external: true
```

`caddy` is my reverse proxy, and I run everything through Cloudflare.

Bind mounts are discouraged in compose, but I love them anyway.

It works perfectly, my dev container setup looks the same.

I don't remember the VM size offhand, but I think it is only 2 vCPU and 8gb or 16gb. As all of the heavy lifting is done by upstream providers it really doesn't take much to run.

There has been some other discussion on this thread about running it as a dockerized web app or similar. That won't be straightforward because of the multiple dependencies, as well as the local storage needed. If you set it up with firebase and a hosted mongodb -- at minimum -- you could get it done, but you're headed torwards a brittle solution.
